plutus-emulator/src/Wallet/Emulator/NodeClient.hs

```diff
@@ -0,0 +1,150 @@
+data ValidatedBlock = ValidatedBlock
+    { vlbValid  :: [Tx]
+    -- ^ The transactions that have been validated in this block.
+    , vlbEvents :: [NodeClientEvent]
+    -- ^ Transaction validation events for the transactions in this block.
+    , vlbRest   :: [Tx]
+    -- ^ The transactions that haven't been validated because the current slot is
+    --   not in their validation interval.
+    }
+
+-- | A pool of transactions which have yet to be validated.
+type TxPool = [Tx]
+
+data NodeClientState = NodeClientState {
+    _chainNewestFirst :: Blockchain, -- ^ The current chain, with the newest transactions first in the list.
+    _txPool           :: TxPool, -- ^ The pool of pending transactions.
+    _index            :: Index.UtxoIndex -- ^ The UTxO index, used for validation.
+} deriving (Show)
+
+makeLenses ''NodeClientState
+
+-- | Validate a block given the current slot and UTxO index, returning the valid
+--   transactions, success/failure events, remaining transactions and the
+--   updated UTxO set.
+validateBlock :: Slot -> Index.UtxoIndex -> [Tx] -> (ValidatedBlock, Index.UtxoIndex)
+validateBlock slot@(Slot s) idx txns =
+    let
+        -- Select those transactions that can be validated in the
+        -- current slot
+        (eligibleTxns, rest) = partition (canValidateNow slot) txns
+
+        -- Validate eligible transactions, updating the UTXO index each time
+        (processed, idx') =
+            flip runState idx $ for eligibleTxns $ \t -> do
+                r <- validateEm slot t
+                pure (t, r)
+
+        -- The new block contains all transaction that were validated
+        -- successfully
+        block = fst <$> filter (isNothing . snd) processed
+
+        -- Also return an `EmulatorEvent` for each transaction that was
+        -- processed
+        nextSlot = Slot (s + 1)
+        events   = SlotAdd nextSlot : (uncurry mkEvent <$> processed)
+
+    in (ValidatedBlock block events rest, idx')
+
+-- | Check whether the given transaction can be validated in the given slot.
+canValidateNow :: Slot -> Tx -> Bool
+canValidateNow slot tx = Interval.member slot (txValidRange tx)
+
+mkEvent :: Tx -> Maybe Index.ValidationError -> NodeClientEvent
+mkEvent t result =
+    case result of
+        Nothing  -> TxnValidate (hashTx t)
+        Just err -> TxnValidationFail (hashTx t) err
+
+-- | Validate a transaction in the current emulator state.
+validateEm :: MonadState Index.UtxoIndex m => Slot -> Tx -> m (Maybe Index.ValidationError)
+validateEm h txn = do
+    idx <- get
+    let result = Index.runValidation (Index.validateTransaction h txn) idx
+    case result of
+        Left e -> pure (Just e)
+        Right idx' -> do
+            _ <- put idx'
+            pure Nothing
+
+class MonadState NodeClientState m => NodeClientAPI m where
+    publishTx :: Tx -> m ()
+    currentSlot :: m Slot
+    eventStream :: Slot -> ConduitT () Block m ()
+    processBlock :: m ValidatedBlock
```

Comment:
As discussed, let's pop these in a separate class for now. And maybe not do `eventStream` just yet.